server/actions_apis.go

```diff
@@ -174,10 +181,18 @@ func (s *RegistryServer) ListApis(ctx context.Context, req *rpc.ListApisRequest)
 	responses := &rpc.ListApisResponse{
 		Apis: apiMessages,
 	}
-	responses.NextPageToken, err = it.GetCursor(len(apiMessages))
+
+	nextToken, err := it.GetCursor(len(apiMessages))
 	if err != nil {
 		return nil, internalError(err)
 	}
+
+	if _, err := it.Next(&api); err == nil {
+		responses.NextPageToken = nextToken
```

Comment:
This change (and the one already made to projects) breaks paginated iteration when filters are used. This only happens for gorm-based queries, and occurs because the gorm queries use a fixed page size (50) to limit the number of records read from the database before applying filters. For large collections, this can cause some pages of a filtered query to return empty arrays, which is allowed by AIP-158 ("The API may return fewer results than the number requested (including zero results), even if not at the end of the collection.")

I agree that we should find a way to return empty page tokens when the end of a collection is reached, but rather than arbitrarily increasing gorm query size or including another layer of iteration in these queries (both of these things could potentially make queries arbitrarily expensive), we might instead look at the iterator returned by client.Run() and (since it is unfiltered), detect the end of the collection by recognizing when that iterator returns nothing.

Reply:
(sorry this wasn't covered by existing tests - I can imagine testing this by creating a largish collection and using filters to select items from the end of the collection, but that also depends on knowing the order of iteration, which we currently do when we specify ids in create calls)